src/JuMP.jl

```diff
@@ -412,40 +412,44 @@ include("quadexpr.jl")
 include("sd.jl")
 
 # handle dictionary of variables
-function registervar(m::Model, varname::Symbol, value)
+function registervar(m::AbstractModel, varname::Symbol, value)
     registerobject(m, varname, value, "A variable or constraint named $varname is already attached to this model. If creating variables programmatically, use the anonymous variable syntax x = @variable(m, [1:N], ...).")
 end
-registervar(m::Model, varname, value) = error("Invalid variable name $varname")
+registervar(m::AbstractModel, varname, value) = error("Invalid variable name $varname")
 
-function registercon(m::Model, conname::Symbol, value)
+function registercon(m::AbstractModel, conname::Symbol, value)
     registerobject(m, conname, value, "A variable or constraint named $conname is already attached to this model. If creating constraints programmatically, use the anonymous constraint syntax con = @constraint(m, ...).")
 end
-registercon(m::Model, conname, value) = error("Invalid constraint name $conname")
+registercon(m::AbstractModel, conname, value) = error("Invalid constraint name $conname")
 
-function registerobject(m::Model, name::Symbol, value, errorstring::String)
-    if haskey(m.objdict, name)
+object_dictionary(m::Model) = m.objdict
```

Comment:
Add a comment that this function needs to be implemented by all `AbstractModels`